docs/source/how-to/migrations/upgrade-to-16.md

```diff
@@ -0,0 +1,70 @@
+# Upgrading your OAuthenticator to version 16.0
+
+The following section describes what to keep in mind when upgrading to OAuthenticator 16.0.
+
+````{important}
+(token_request)=
+
+```{rubric} The token request
+```
+Whenever _token request_ is used, it refers to the HTTP request exchanging the [oauth code](https://www.rfc-editor.org/rfc/rfc6749#section-1.3.1) for the [access token](https://www.rfc-editor.org/rfc/rfc6749#section-1.4).
+
+This request is sent to the [`token_url`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.oauth2.html#oauthenticator.oauth2.OAuthenticator.token_url) in [`get_token_info()`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.oauth2.html#oauthenticator.oauth2.OAuthenticator.get_token_info) method.
+
+(userdata_request)=
+
+```{rubric} The userdata request
+```
+Whenever _userdata request_ is used, it refers to the HTTP request that's exchanging the the `access token` to get the `userdata`.
+
+This request is sent to [`userdata_url`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.oauth2.html#oauthenticator.oauth2.OAuthenticator.userdata_url) in [`token_to_user`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.oauth2.html#oauthenticator.oauth2.OAuthenticator.get_token_info).
+````
+
+## Deprecations
+
+1. `username_key` replaced by `username_claim` in _all oauthenticators_
+
+- [`GenericOAuthenticator.username_key`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.generic.html#oauthenticator.generic.GenericOAuthenticator.username_key) is deprecated and replaced by [`username_claim`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.oauth2.html#oauthenticator.oauth2.OAuthenticator.username_claim).
+
+- [`Auth0OAuthenticator.username_key`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.auth0.html#oauthenticator.auth0.Auth0OAuthenticator.username_key) is deprecated and replaced by [`username_claim`](https://oauthenticator.readthedocs.io/en/latest/reference/api/gen/oauthenticator.oauth2.html#oauthenticator.oauth2.OAuthenticator.username_claim).
```

Comment:
Some indentation matters here and elsewhere, where the content of `1.` isn't indented.